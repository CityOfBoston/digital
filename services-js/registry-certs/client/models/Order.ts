import { observable, computed, action } from 'mobx';
import makeShippingValidator, {
  ShippingValidator,
} from '../../lib/validators/ShippingValidator';
import makePaymentValidator, {
  PaymentValidator,
} from '../../lib/validators/PaymentValidator';

// Session-storage based container for keeping track of ordering info. Only add
// values here that should be in session storage.
export interface OrderInfo {
  storeContactAndShipping: boolean;
  storeBilling: boolean;

  contactName: string;
  contactEmail: string;
  contactPhone: string;

  shippingName: string;
  shippingCompanyName: string;
  shippingAddress1: string;
  shippingAddress2: string;
  shippingCity: string;
  shippingState: string;
  shippingZip: string;

  cardholderName: string;
  cardLast4: string;
  cardToken: string | null;
  cardFunding: 'credit' | 'debit' | 'prepaid' | 'unknown';

  billingAddressSameAsShippingAddress: boolean;

  billingAddress1: string;
  billingAddress2: string;
  billingCity: string;
  billingState: string;
  billingZip: string;
}

export default class Order {
  @observable info: OrderInfo = null as any;

  idempotencyKey: string | null = null;

  // Set to true if there's a network operation related to the order going on,
  // such as tokenizing the card with Stripe or submitting the order to the
  // backend.
  @observable processing: boolean = false;

  updateStorageDisposer: Function | null = null;

  readonly localStorageAvailable: boolean;

  constructor(info: OrderInfo | null = null, localStorageAvailable = true) {
    this.localStorageAvailable = localStorageAvailable;

    this.info = info || {
      storeContactAndShipping: false,
      storeBilling: false,

      contactName: '',
      contactEmail: '',
      contactPhone: '',

      shippingName: '',
      shippingCompanyName: '',
      shippingAddress1: '',
      shippingAddress2: '',
      shippingCity: '',
      shippingState: 'MA',
      shippingZip: '',

      cardholderName: '',
      cardLast4: '',
      cardToken: null,
      cardFunding: 'unknown',

      billingAddressSameAsShippingAddress: true,

      billingAddress1: '',
      billingAddress2: '',
      billingCity: '',
      billingState: 'MA',
      billingZip: '',
    };
  }

  @action
  /**
   * Modifies the info data by updating it with the provided data. Really just a
   * type-safe wrapper around Object.assign.
   */
  updateInfo(partialInfo: Partial<OrderInfo>) {
    // TIL: TypeScriptâ€™s typing of Object.assign does not guarantee that the
    // first parameter (which gets mutated) maintains its type.
    Object.assign(this.info, partialInfo);
  }

  @computed
  get shippingValidator(): ShippingValidator {
    const validator = makeShippingValidator(this.info);
    validator.check();
    return validator;
  }

  @computed
  get shippingIsComplete(): boolean {
    return this.shippingValidator.passes();
  }

  @computed
  get shippingErrors(): { [key: string]: Array<string> } {
    return this.shippingValidator.errors.all();
  }

  @computed
  get paymentValidator(): PaymentValidator {
    const validator = makePaymentValidator(this.info);
    validator.check();
    return validator;
  }

  @computed
  get paymentIsComplete(): boolean {
    return this.paymentValidator.passes() && this.info.cardToken !== null;
  }

  @computed
  get paymentErrors(): { [key: string]: Array<string> } {
    return this.paymentValidator.errors.all();
  }

  @computed
  get billingAddress1(): string {
    return this.info.billingAddressSameAsShippingAddress
      ? this.info.shippingAddress1
      : this.info.billingAddress1;
  }

  @computed
  get billingAddress2(): string {
    return this.info.billingAddressSameAsShippingAddress
      ? this.info.shippingAddress2
      : this.info.billingAddress2;
  }

  @computed
  get billingCity(): string {
    return this.info.billingAddressSameAsShippingAddress
      ? this.info.shippingCity
      : this.info.billingCity;
  }

  @computed
  get billingState(): string {
    return this.info.billingAddressSameAsShippingAddress
      ? this.info.shippingState
      : this.info.billingState;
  }

  @computed
  get billingZip(): string {
    return this.info.billingAddressSameAsShippingAddress
      ? this.info.shippingZip
      : this.info.billingZip;
  }

  @action
  resetCard() {
    this.info.cardToken = null;
    this.info.cardFunding = 'credit';
    this.info.cardLast4 = '';
  }

  // We use an idempotency key to prevent double-clicks on submit from
  // generating mulitple orders (though disabling submit during submission helps
  // as well).
  regenerateIdempotencyKey() {
    this.idempotencyKey = Math.random()
      .toString(36)
      .substring(2, 9);
  }
}
