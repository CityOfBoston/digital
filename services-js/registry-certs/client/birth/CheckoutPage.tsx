// Wrapper controller for the separate pages along the checkout flow

import React from 'react';
import { observer } from 'mobx-react';
import Router from 'next/router';

import { PageDependencies, GetInitialProps } from '../../pages/_app';
import Order, { OrderInfo } from '../models/Order';
import { BIRTH_CERTIFICATE_COST } from '../../lib/costs';

import ShippingContent from '../common/checkout/ShippingContent';
import PaymentContent from '../common/checkout/PaymentContent';
import ReviewContent from '../common/checkout/ReviewContent';
import CheckoutPageLayout from '../common/checkout/CheckoutPageLayout';
import BirthConfirmationContent from './BirthConfirmationContent';

type PageInfo =
  | {
      page: 'shipping';
    }
  | {
      page: 'payment';
    }
  | {
      page: 'review';
    }
  | {
      page: 'confirmation';
      orderId: string;
      contactEmail: string;
      stepCount: number;
    };

interface InitialProps {
  info: PageInfo;
}

export type PageDependenciesProps = Pick<
  PageDependencies,
  | 'birthCertificateRequest'
  | 'siteAnalytics'
  | 'orderProvider'
  | 'checkoutDao'
  | 'stripe'
>;

interface Props extends InitialProps, PageDependenciesProps {
  orderForTest?: Order;
}

type State = {
  /**
   * This will be null on the server and during the first client render.
   */
  order: Order | null;
};

@observer
/**
 * This is currently copied over from death’s checkout page, though some
 * elements could potentially be generalized out from them.
 */
export default class BirthCheckoutPage extends React.Component<Props, State> {
  static getInitialProps: GetInitialProps<InitialProps, 'query'> = ({
    query,
  }) => {
    let info: PageInfo;

    const page = query.page || '';

    switch (page) {
      case '':
      case 'shipping':
        info = { page: 'shipping' };
        break;
      case 'payment':
        info = { page: 'payment' };
        break;
      case 'review':
        info = { page: 'review' };
        break;
      case 'confirmation':
        info = {
          page: 'confirmation',
          orderId: query.orderId || '',
          contactEmail: query.contactEmail || '',
          stepCount: parseInt(query.stepCount || '8'),
        };
        break;
      default:
        info = { page: 'shipping' };
        break;
    }

    return { info };
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      order: props.orderForTest || null,
    };
  }

  async componentDidMount() {
    this.reportCheckoutStep(this.props);

    const { orderProvider, siteAnalytics } = this.props;

    // If any of the question steps are not complete (i.e. user backed up,
    // changed an answer, and used the browser “forward” button), return to the
    // beginning of the questions flow.
    if (this.props.birthCertificateRequest.questionStepsComplete === false) {
      // todo: improve this experience
      // todo: behavior is duplicated in ReviewRequestPage.tsx

      siteAnalytics.sendEvent('question results lost', {
        category: 'Birth',
        label: `checkout: ${this.props.info.page}`,
      });

      return Router.push('/birth');
    }

    if (this.state.order) {
      return;
    }

    // We won’t have an Order until we’re mounted in the browser because it’s
    // dependent on sessionStorage / localStorage data.
    const order = await orderProvider.get();
    await new Promise(resolve => this.setState({ order }, resolve));
  }

  componentWillReceiveProps(newProps: Props) {
    if (newProps.info.page !== this.props.info.page) {
      this.reportCheckoutStep(newProps);
    }
  }

  reportCheckoutStep({ info, siteAnalytics }: Props) {
    let checkoutStep: number | null = null;
    switch (info.page) {
      case 'shipping':
        checkoutStep = 1;
        break;
      case 'payment':
        checkoutStep = 2;
        break;
      case 'review':
        checkoutStep = 3;
        break;
    }

    if (checkoutStep) {
      this.birthCertificateProduct();

      siteAnalytics.setProductAction('checkout', { step: checkoutStep });
    }
  }

  advanceToPayment = async (shippingInfo: Partial<OrderInfo>) => {
    const { order } = this.state;

    if (!order) {
      return;
    }

    order.updateInfo(shippingInfo);

    await Router.push('/birth/checkout?page=payment');

    window.scroll(0, 0);
  };

  advanceToReview = async (
    cardElement: stripe.elements.Element | null,
    billingInfo: Partial<OrderInfo>
  ) => {
    const { checkoutDao } = this.props;
    const { order } = this.state;

    if (!order) {
      return;
    }

    order.updateInfo(billingInfo);

    // This may throw, in which case the payment page will catch it and display
    // the error.
    if (cardElement) {
      await checkoutDao.tokenizeCard(order, cardElement);
    }

    await Router.push('/birth/checkout?page=review');

    window.scroll(0, 0);
  };

  /**
   * Submits the order.
   *
   * Will throw exceptions if things didn’t go well.
   */
  submitOrder = async () => {
    const {
      checkoutDao,
      birthCertificateRequest,
      siteAnalytics,
      orderProvider,
    } = this.props;

    const { order } = this.state;

    if (!order) {
      return;
    }

    const orderId = await checkoutDao.submitBirthCertificateRequest(
      birthCertificateRequest,
      order
    );

    this.birthCertificateProduct();

    siteAnalytics.setProductAction('purchase', {
      id: orderId,
      revenue: birthCertificateRequest.quantity * BIRTH_CERTIFICATE_COST / 100,
    });

    siteAnalytics.sendEvent('click', {
      category: 'Birth',
      label: 'submit order',
    });

    siteAnalytics.sendEvent('ship to city', {
      category: 'Birth',
      label: `${order.info.shippingCity}, ${order.info.shippingState}`,
    });

    siteAnalytics.sendEvent('ship to state', {
      category: 'Birth',
      label: order.info.shippingState,
    });

    siteAnalytics.sendEvent('place order', {
      category: 'Birth',
      label: 'certificate quantity',
      value: birthCertificateRequest.quantity,
    });

    const stepCount = birthCertificateRequest.steps.length;

    birthCertificateRequest.clearBirthCertificateRequest();
    orderProvider.clear();

    this.setState(
      {
        order: await orderProvider.get(),
      },
      async () => {
        await Router.push(
          `/birth/checkout?page=confirmation&orderId=${encodeURIComponent(
            orderId
          )}&contactEmail=${encodeURIComponent(
            order.info.contactEmail
          )}&stepCount=${stepCount}`,
          '/birth/checkout?page=confirmation'
        );

        window.scroll(0, 0);
      }
    );
  };

  birthCertificateProduct() {
    this.props.siteAnalytics.addProduct(
      '0',
      'Birth certificate',
      'Birth certificate',
      this.props.birthCertificateRequest.quantity,
      BIRTH_CERTIFICATE_COST / 100
    );
  }

  render() {
    const { info, birthCertificateRequest, stripe } = this.props;
    const { order } = this.state;

    const progressSteps = birthCertificateRequest.steps.length;

    // This happens during server side rendering
    if (!order) {
      return <CheckoutPageLayout certificateType="birth" />;
    }

    switch (info.page) {
      case 'shipping':
        return (
          <ShippingContent
            certificateType="birth"
            birthCertificateRequest={birthCertificateRequest}
            order={order}
            submit={this.advanceToPayment}
            progress={{
              currentStep:
                birthCertificateRequest.steps.indexOf('shippingInformation') +
                1,
              currentStepCompleted: false,
              totalSteps: progressSteps,
            }}
          />
        );

      case 'payment':
        return (
          <PaymentContent
            certificateType="birth"
            stripe={stripe}
            birthCertificateRequest={birthCertificateRequest}
            order={order}
            submit={this.advanceToReview}
            progress={{
              currentStep:
                birthCertificateRequest.steps.indexOf('billingInformation') + 1,
              currentStepCompleted: false,
              totalSteps: progressSteps,
            }}
          />
        );

      case 'review':
        return (
          <ReviewContent
            certificateType="birth"
            birthCertificateRequest={birthCertificateRequest}
            order={order}
            submit={this.submitOrder}
            progress={{
              currentStep:
                birthCertificateRequest.steps.indexOf('submitRequest') + 1,
              currentStepCompleted: false,
              totalSteps: progressSteps,
            }}
          />
        );

      case 'confirmation':
        return (
          <BirthConfirmationContent
            contactEmail={info.contactEmail}
            orderId={info.orderId}
            stepCount={info.stepCount}
          />
        );
    }
  }
}
