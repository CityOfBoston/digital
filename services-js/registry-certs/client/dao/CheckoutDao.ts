import { action, runInAction } from 'mobx';

import { FetchGraphql } from '@cityofboston/next-client-common';

import DeathCertificateCart from '../store/DeathCertificateCart';
import BirthCertificateRequest from '../store/BirthCertificateRequest';
import MarriageCertificateRequest from '../store/MarriageCertificateRequest';

import Order from '../models/Order';

import submitDeathCertificateOrder from '../queries/submit-death-certificate-order';
import submitBirthCertificateOrder from '../queries/submit-birth-certificate-order';
import submitMarriageCertificateOrder from '../queries/submit-marriage-certificate-order';

import { OrderErrorCause } from '../queries/graphql-types';
import {
  DeathCertificateOrderResult,
  BirthCertificateOrderResult,
  MarriageCertificateOrderResult,
} from '../types';

export class SubmissionError extends Error {
  public readonly cause: OrderErrorCause;

  constructor(message: string, cause: OrderErrorCause) {
    super(message);
    this.cause = cause;
  }
}

export default class CheckoutDao {
  fetchGraphql: FetchGraphql;
  stripe: stripe.Stripe | null;

  constructor(fetchGraphql: FetchGraphql, stripe: stripe.Stripe | null) {
    this.fetchGraphql = fetchGraphql;
    this.stripe = stripe;
  }

  /**
   * Given a StripeElement object, tokenizes the card info with Stripe and, if
   * successful, populates the cardToken and cardLast4 values in Order’s info.
   *
   * We rely on Stripe’s form validation to catch straightforward errors (such
   * as a card number that doesn’t match the formula) so we’re fairly sure that
   * tokenization will succeed so we don’t need to handle the errors very
   * delicately.
   *
   * Sets Order#processing to true while the API call is outstanding
   *
   * @throws An Error if things did not work. Reports to Rollbar so that callers
   * don’t have to.
   */
  @action
  async tokenizeCard(
    order: Order,
    cardElement: stripe.elements.Element | null
  ): Promise<void> {
    const { stripe } = this;

    if (!stripe || !cardElement) {
      throw new Error(
        'tokenizeCard called without Stripe and/or StripeElement'
      );
    }

    try {
      order.processing = true;

      const tokenResult = await stripe.createToken(cardElement, {
        name: order.info.cardholderName,
        address_line1: order.billingAddress1,
        address_line2: order.billingAddress2,
        address_city: order.billingCity,
        address_state: order.billingState,
        address_zip: order.billingZip,
        address_country: 'us',
      });

      if (tokenResult.error) {
        throw new Error(tokenResult.error.message);
      } else {
        const { token } = tokenResult;

        if (!token) {
          throw new Error('Token was not in token result');
        }

        const { card } = token;

        if (!card) {
          throw new Error('Card was not in token result');
        }

        runInAction(() => {
          order.info.cardToken = token.id;
          order.info.cardFunding = card.funding;
          order.info.cardLast4 = card.last4;
        });
      }
    } catch (err) {
      if ((window as any).Rollbar && !err._reportedException) {
        (window as any).Rollbar.error(err);
        err._reportedException = true;
      }

      throw err;
    } finally {
      runInAction(() => {
        order.processing = false;
      });
    }
  }

  /**
   * Submits the given cart’s worth of certificates with the order data
   * (address, &c.).
   *
   * Sets Order#processing to true while the API call is outstanding
   *
   * @returns The order ID on success.
   * @throws Error or SubmissionError objects. Reports errors to Rollbar.
   */
  submitDeathCertificateCart(
    cart: DeathCertificateCart,
    order: Order
  ): Promise<string> {
    const orderPromise = submitDeathCertificateOrder(
      this.fetchGraphql,
      cart,
      order
    );

    return this.handleOrder(order, orderPromise);
  }

  /**
   * Submits a birth certificate request with the order data (address, &c.).
   *
   * Sets Order#processing to true while the API call is outstanding
   *
   * @returns The order ID on success.
   * @throws Error or SubmissionError objects. Reports errors to Rollbar.
   */
  submitBirthCertificateRequest(
    birthCertificateRequest: BirthCertificateRequest,
    order: Order
  ): Promise<string> {
    const orderPromise = submitBirthCertificateOrder(
      this.fetchGraphql,
      birthCertificateRequest,
      order
    );

    return this.handleOrder(order, orderPromise);
  }

  /**
   * Copied from submitBirthCertificateRequest() above:
   *
   * Submits a marriage certificate request with the order data (address, &c.).
   *
   * Sets Order#processing to true while the API call is outstanding
   *
   * @returns The order ID on success.
   * @throws Error or SubmissionError objects. Reports errors to Rollbar.
   */
  submitMarriageCertificateRequest(
    marriageCertificateRequest: MarriageCertificateRequest,
    order: Order
  ): Promise<string> {
    const orderPromise = submitMarriageCertificateOrder(
      this.fetchGraphql,
      marriageCertificateRequest,
      order
    );

    return this.handleOrder(order, orderPromise);
  }

  @action
  private async handleOrder(
    order: Order,
    orderPromise: Promise<
      | DeathCertificateOrderResult
      | BirthCertificateOrderResult
      | MarriageCertificateOrderResult
    >
  ): Promise<string> {
    try {
      order.processing = true;

      let orderResult:
        | DeathCertificateOrderResult
        | BirthCertificateOrderResult
        | MarriageCertificateOrderResult;

      try {
        orderResult = await orderPromise;
      } catch (err) {
        // These errors will be network sorts of errors.
        if ((window as any).Rollbar && !err._reportedException) {
          (window as any).Rollbar.error(err);
          err._reportedException = true;
        }

        throw err;
      }

      if (orderResult.order) {
        return orderResult.order.id;
      } else if (orderResult.error) {
        // We don't need this to be reported to Rollbar. If it’s an internal
        // error, the server would have reported it, and if it’s a user error
        // (card information) we don’t care.
        throw new SubmissionError(
          orderResult.error.message,
          orderResult.error.cause
        );
      } else {
        // This won’t happen
        throw new Error('Result did not have an order or error');
      }
    } finally {
      runInAction(() => {
        order.processing = false;
      });
    }
  }
}
