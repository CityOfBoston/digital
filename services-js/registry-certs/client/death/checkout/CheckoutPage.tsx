// Wrapper controller for the separate pages along the checkout flow

import React from 'react';
import { observer } from 'mobx-react';
import Router from 'next/router';

import { getParam } from '@cityofboston/next-client-common';

import { PageDependencies, GetInitialProps } from '../../../pages/_app';
import Order, { OrderInfo } from '../../models/Order';

import { CERTIFICATE_COST } from '../../../lib/costs';

import ShippingContent from '../../common/checkout/ShippingContent';
import PaymentContent from '../../common/checkout/PaymentContent';
import ReviewContent from '../../common/checkout/ReviewContent';
import ConfirmationContent from './ConfirmationContent';
import CheckoutPageLayout from '../../common/checkout/CheckoutPageLayout';

const DEATH_CERTIFICATE_COST = CERTIFICATE_COST.DEATH;

type PageInfo =
  | {
      page: 'shipping';
    }
  | {
      page: 'payment';
    }
  | {
      page: 'review';
    }
  | {
      page: 'confirmation';
      orderId: string;
      contactEmail: string;
    };

interface InitialProps {
  info: PageInfo;
}

export type PageDependenciesProps = Pick<
  PageDependencies,
  | 'deathCertificateCart'
  | 'siteAnalytics'
  | 'orderProvider'
  | 'checkoutDao'
  | 'stripe'
>;

interface Props extends InitialProps, PageDependenciesProps {
  orderForTest?: Order;
}

type State = {
  /**
   * This will be null on the server and during the first client render.
   */
  order: Order | null;
};

@observer
export default class CheckoutPageController extends React.Component<
  Props,
  State
> {
  state: State = {
    order: this.props.orderForTest || null,
  };

  static getInitialProps: GetInitialProps<InitialProps, 'query'> = ({
    query,
  }) => {
    let info: PageInfo;

    const page = query.page || '';

    switch (page) {
      case '':
      case 'shipping':
        info = { page: 'shipping' };
        break;
      case 'payment':
        info = { page: 'payment' };
        break;
      case 'review':
        info = { page: 'review' };
        break;
      case 'confirmation':
        info = {
          page: 'confirmation',
          orderId: getParam(query.orderId, ''),
          contactEmail: getParam(query.contactEmail, ''),
        };
        break;
      default:
        info = { page: 'shipping' };
        break;
    }

    return { info };
  };

  async componentDidMount() {
    this.reportCheckoutStep(this.props);

    const { orderProvider } = this.props;

    // We won’t have an Order until we’re mounted in the browser because it’s
    // dependent on sessionStorage / localStorage data.
    const order = await orderProvider.get();
    await new Promise(resolve => this.setState({ order }, resolve));
  }

  componentWillReceiveProps(newProps: Props) {
    if (newProps.info.page !== this.props.info.page) {
      this.reportCheckoutStep(newProps);
    }
  }

  reportCheckoutStep({ info, deathCertificateCart, siteAnalytics }: Props) {
    let checkoutStep: number | null = null;
    switch (info.page) {
      case 'shipping':
        checkoutStep = 1;
        break;
      case 'payment':
        checkoutStep = 2;
        break;
      case 'review':
        checkoutStep = 3;
        break;
    }

    if (checkoutStep) {
      deathCertificateCart.trackCartItems();
      siteAnalytics.setProductAction('checkout', { step: checkoutStep });
    }
  }

  advanceToPayment = async (shippingInfo: Partial<OrderInfo>) => {
    const { order } = this.state;

    if (!order) {
      return;
    }

    order.updateInfo(shippingInfo);

    await Router.push('/death/checkout?page=payment');

    window.scroll(0, 0);
  };

  advanceToReview = async (
    cardElement: stripe.elements.Element | null,
    billingInfo: Partial<OrderInfo>
  ) => {
    const { checkoutDao } = this.props;
    const { order } = this.state;

    if (!order) {
      return;
    }

    order.updateInfo(billingInfo);

    // This may throw, in which case the payment page will catch it and display
    // the error.
    if (cardElement) {
      await checkoutDao.tokenizeCard(order, cardElement);
    }

    await Router.push('/death/checkout?page=review');

    window.scroll(0, 0);
  };

  /**
   * Submits the order.
   *
   * Will throw exceptions if things didn’t go well.
   */
  submitOrder = async () => {
    const {
      deathCertificateCart,
      checkoutDao,
      siteAnalytics,
      orderProvider,
    } = this.props;

    const { order } = this.state;

    if (!order) {
      return;
    }

    const orderId = await checkoutDao.submitDeathCertificateCart(
      deathCertificateCart,
      order
    );

    deathCertificateCart.trackCartItems();

    siteAnalytics.setProductAction('purchase', {
      id: orderId,
      revenue: (deathCertificateCart.size * DEATH_CERTIFICATE_COST) / 100,
    });

    siteAnalytics.sendEvent('click', {
      category: 'UX',
      label: 'submit order',
    });

    deathCertificateCart.clear();
    orderProvider.clear();

    this.setState({
      order: await orderProvider.get(),
    });

    await Router.push(
      `/death/checkout?page=confirmation&orderId=${encodeURIComponent(
        orderId
      )}&contactEmail=${encodeURIComponent(order.info.contactEmail)}`,
      '/death/checkout?page=confirmation'
    );

    window.scroll(0, 0);
  };

  render() {
    const { info, deathCertificateCart, stripe } = this.props;
    const { order } = this.state;

    // This happens during server side rendering
    if (!order) {
      return <CheckoutPageLayout certificateType="death" />;
    }

    switch (info.page) {
      case 'shipping':
        return (
          <ShippingContent
            certificateType="death"
            deathCertificateCart={deathCertificateCart}
            order={order}
            submit={this.advanceToPayment}
          />
        );

      case 'payment':
        return (
          <PaymentContent
            certificateType="death"
            stripe={stripe}
            deathCertificateCart={deathCertificateCart}
            order={order}
            submit={this.advanceToReview}
          />
        );

      case 'review':
        return (
          <ReviewContent
            certificateType="death"
            deathCertificateCart={deathCertificateCart}
            order={order}
            submit={this.submitOrder}
          />
        );

      case 'confirmation':
        return (
          <ConfirmationContent
            orderId={info.orderId}
            contactEmail={info.contactEmail}
            cart={deathCertificateCart}
          />
        );
    }
  }
}
