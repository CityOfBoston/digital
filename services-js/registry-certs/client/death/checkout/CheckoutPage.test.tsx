import Router from 'next/router';

import { GaSiteAnalytics } from '@cityofboston/next-client-common';

import CheckoutPage, { PageDependenciesProps } from './CheckoutPage';
import CheckoutDao, { SubmissionError } from '../../dao/CheckoutDao';
import OrderProvider from '../../store/OrderProvider';
import DeathCertificateCart from '../../store/DeathCertificateCart';
import { OrderErrorCause } from '../../queries/graphql-types';
import Order from '../../models/Order';

jest.mock('next/router');
jest.mock('../../dao/CheckoutDao');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('getInitialProps', () => {
  it('treats no page query param as shipping', async () => {
    const initialProps = await CheckoutPage.getInitialProps({ query: {} }, {});
    expect(initialProps.info.page).toEqual('shipping');
  });

  it('process payment on server', async () => {
    const initialProps = await CheckoutPage.getInitialProps(
      { query: { page: 'payment' } },
      {}
    );

    expect(initialProps.info.page).toEqual('payment');
  });

  it('treats unknown page query param as shipping on client', async () => {
    const initialProps = await CheckoutPage.getInitialProps(
      {
        query: { page: 'not-a-real-page' },
      },
      {}
    );

    expect(initialProps.info.page).toEqual('shipping');
  });

  it('passes confirm props along', async () => {
    const initialProps = await CheckoutPage.getInitialProps(
      {
        query: {
          page: 'confirmation',
          orderId: '123-456-7',
          contactEmail: 'ttoe@squirrelzone.net',
        },
      },
      {}
    );

    expect(initialProps.info).toEqual({
      page: 'confirmation',
      orderId: '123-456-7',
      contactEmail: 'ttoe@squirrelzone.net',
    });
  });
});

describe('rendering', () => {
  let pageDependenciesProps: PageDependenciesProps;

  beforeEach(() => {
    pageDependenciesProps = {
      deathCertificateCart: new DeathCertificateCart(),
      checkoutDao: {} as any,
      orderProvider: new OrderProvider(),
      siteAnalytics: new GaSiteAnalytics(),
      stripe: null,
    };
  });

  it('renders shipping', () => {
    const page = new CheckoutPage({
      info: { page: 'shipping' },
      ...pageDependenciesProps,
    });

    page.state = {
      order: new Order(),
    };

    expect(page.render()).toMatchSnapshot();
  });

  it('renders payment', () => {
    const page = new CheckoutPage({
      info: { page: 'payment' },
      ...pageDependenciesProps,
    });

    page.state = {
      order: new Order(),
    };

    expect(page.render()).toMatchSnapshot();
  });

  it('renders review', () => {
    const page = new CheckoutPage({
      info: { page: 'review' },
      ...pageDependenciesProps,
    });

    page.state = {
      order: new Order(),
    };

    expect(page.render()).toMatchSnapshot();
  });

  it('renders confirmation', () => {
    const page = new CheckoutPage({
      info: {
        page: 'confirmation',
        orderId: '123-456-7',
        contactEmail: 'ttoe@squirrelzone.net',
      },
      ...pageDependenciesProps,
    });

    page.state = {
      order: new Order(),
    };

    expect(page.render()).toMatchSnapshot();
  });
});

describe('operations', () => {
  let checkoutDao: CheckoutDao;
  let orderProvider: OrderProvider;
  let component: CheckoutPage;
  let scrollSpy;

  beforeEach(async () => {
    scrollSpy = jest.spyOn(window, 'scroll').mockImplementation(() => {});

    orderProvider = new OrderProvider();
    orderProvider.attach(null, null);
    checkoutDao = new CheckoutDao(null as any, null);

    // page doesn't really matter for this
    component = new CheckoutPage({
      deathCertificateCart: new DeathCertificateCart(),
      siteAnalytics: new GaSiteAnalytics(),
      stripe: null,

      info: { page: 'shipping' },
      orderProvider,
      checkoutDao,
    });

    component.state = {
      order: await orderProvider.get(),
    };

    // Keeps us from setting state in the order succes case, since it will fail
    // (the component isnâ€™t actually mounted)
    component.setState = jest.fn();
  });

  afterEach(() => {
    scrollSpy.mockRestore();
  });

  describe('advanceToPayment', () => {
    it('routes to next stage', () => {
      component.advanceToPayment({} as any);
      expect(Router.push).toHaveBeenCalled();
    });
  });

  describe('advanceToReview', () => {
    it('routes to next stage if tokenization is successful', async () => {
      (checkoutDao.tokenizeCard as jest.Mock).mockReturnValue(
        Promise.resolve()
      );

      await component.advanceToReview(null, {} as any);
      expect(Router.push).toHaveBeenCalled();
    });

    it('does not route if tokenization is unsuccessful', async () => {
      (checkoutDao.tokenizeCard as jest.Mock).mockReturnValue(
        Promise.reject(new Error('tokenization failed'))
      );

      await expect(
        component.advanceToReview({} as any, {} as any)
      ).rejects.toMatchObject({
        message: 'tokenization failed',
      });
      expect(Router.push).not.toHaveBeenCalled();
    });
  });

  describe('submitOrder', () => {
    it('redirects on success', async () => {
      (checkoutDao.submitDeathCertificateCart as jest.Mock).mockReturnValue(
        Promise.resolve('test-order-id')
      );
      await component.submitOrder();
      expect(Router.push).toHaveBeenCalled();
    });

    it('stays on the same page on failure', async () => {
      (checkoutDao.submitDeathCertificateCart as jest.Mock).mockReturnValue(
        Promise.reject(
          new SubmissionError(
            'The card is expired',
            OrderErrorCause.USER_PAYMENT
          )
        )
      );

      await expect(component.submitOrder()).rejects.toBeInstanceOf(
        SubmissionError
      );

      expect(Router.push).not.toHaveBeenCalled();
    });
  });
});
