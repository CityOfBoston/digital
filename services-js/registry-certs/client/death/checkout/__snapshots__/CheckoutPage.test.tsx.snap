// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders confirmation 1`] = `
<ConfirmationContent
  cart={
    DeathCertificateCart {
      "localStorageDisposer": null,
      "siteAnalytics": null,
    }
  }
  contactEmail="ttoe@squirrelzone.net"
  orderId="123-456-7"
/>
`;

exports[`rendering renders payment 1`] = `
<PaymentContent
  certificateType="death"
  deathCertificateCart={
    DeathCertificateCart {
      "localStorageDisposer": null,
      "siteAnalytics": null,
    }
  }
  order={
    Order {
      "idempotencyKey": null,
      "info": Object {
        "billingAddress1": "",
        "billingAddress2": "",
        "billingAddressSameAsShippingAddress": true,
        "billingCity": "",
        "billingState": "MA",
        "billingZip": "",
        "cardFunding": "unknown",
        "cardLast4": "",
        "cardToken": null,
        "cardholderName": "",
        "contactEmail": "",
        "contactName": "",
        "contactPhone": "",
        "shippingAddress1": "",
        "shippingAddress2": "",
        "shippingCity": "",
        "shippingCompanyName": "",
        "shippingName": "",
        "shippingState": "MA",
        "shippingZip": "",
        "storeBilling": false,
        "storeContactAndShipping": false,
      },
      "localStorageAvailable": true,
      "processing": false,
      "updateStorageDisposer": null,
    }
  }
  stripe={null}
  submit={[Function]}
/>
`;

exports[`rendering renders review 1`] = `
<ReviewContent
  certificateType="death"
  deathCertificateCart={
    DeathCertificateCart {
      "localStorageDisposer": null,
      "siteAnalytics": null,
    }
  }
  order={
    Order {
      "idempotencyKey": null,
      "info": Object {
        "billingAddress1": "",
        "billingAddress2": "",
        "billingAddressSameAsShippingAddress": true,
        "billingCity": "",
        "billingState": "MA",
        "billingZip": "",
        "cardFunding": "unknown",
        "cardLast4": "",
        "cardToken": null,
        "cardholderName": "",
        "contactEmail": "",
        "contactName": "",
        "contactPhone": "",
        "shippingAddress1": "",
        "shippingAddress2": "",
        "shippingCity": "",
        "shippingCompanyName": "",
        "shippingName": "",
        "shippingState": "MA",
        "shippingZip": "",
        "storeBilling": false,
        "storeContactAndShipping": false,
      },
      "localStorageAvailable": true,
      "processing": false,
      "updateStorageDisposer": null,
    }
  }
  submit={[Function]}
/>
`;

exports[`rendering renders shipping 1`] = `
<ShippingContent
  certificateType="death"
  deathCertificateCart={
    DeathCertificateCart {
      "localStorageDisposer": null,
      "siteAnalytics": null,
    }
  }
  order={
    Order {
      "idempotencyKey": null,
      "info": Object {
        "billingAddress1": "",
        "billingAddress2": "",
        "billingAddressSameAsShippingAddress": true,
        "billingCity": "",
        "billingState": "MA",
        "billingZip": "",
        "cardFunding": "unknown",
        "cardLast4": "",
        "cardToken": null,
        "cardholderName": "",
        "contactEmail": "",
        "contactName": "",
        "contactPhone": "",
        "shippingAddress1": "",
        "shippingAddress2": "",
        "shippingCity": "",
        "shippingCompanyName": "",
        "shippingName": "",
        "shippingState": "MA",
        "shippingZip": "",
        "storeBilling": false,
        "storeContactAndShipping": false,
      },
      "localStorageAvailable": true,
      "processing": false,
      "updateStorageDisposer": null,
    }
  }
  submit={[Function]}
/>
`;
